### ISSUE TITLE
Note content retrieval very slow and no progress indicator

### ISSUE THREAD

**Issue by robert7** [d1vanov/quentier#206]
I very often (mainly after program startup) get following situation:
![screen_20180722_06](https://user-images.githubusercontent.com/1135218/43048581-27306d38-8dea-11e8-960b-5ae4a07b6df6.png)

I select some note and it displays "Please select some existing note or create new one" - which is not correct - as it is existing note. Note then really displays - but sometimes it takes like 10-15 seconds (which I consider quite slow). There is no indication, that some background process in running (and seemingly it is).

So when I click like 5 times on 5 notes - I get 5 tabs open with blank title - the notes then load later - but it look weird. So at least some progress info would be appreciated. And check why loading takes really so long (I have quite fast computer with SSD disk).

**Comment by d1vanov** [comment]
Are you running sync when that happens? It could explain the slowness

**Comment by robert7** [comment]
No. Sync is not running. And it is really **very slow**. Like I display a long list of notes (I have ~4500). Then I scroll and randomly open notes. It takes like 10-15 second to open most notes.

**Comment by robert7** [comment]
After the App is running for a while (like 10 minutes) then displaying notes becomes instant - like if all would be loaded into memory. But after app restart it is again very slow.

**Comment by d1vanov** [comment]
If you select the same notes after ten minutes, they might indeed come from the cache. There are two levels of cache (not accounting for SQLite's own momety cache) - outside local storage classes and within them. I'm writing this from my phone so can't give precise code references now.

It might be that the local storage is busy serving the note model which lists all account's notes in chunks of fixed size. It occurs after the app's startup. Look at `src/models/NoteModel.cpp`, understand how it subscribes to notes. Try to limit the model - make it load only the first chunk from the local storage and see if it makes any difference. If it won't, it might be that the problem is related to searching for a particular nore in the local storage - probably some index is missing in the database and slow linear search is employed.

**Comment by robert7** [comment]
And also the search is pretty slow and even if I don't see the result set count. For some cases it for sure misses some notes. E.g. nixnote finds 15 notes and Quentier displays 6 (then after 1 minute (!) Quentier displays 13 notes).

**Comment by d1vanov** [comment]
I think I know why the searching feels so slow: `NoteFilterModel` contains `m_noteLocalUids` which is `QStringList` so calling `contains` on this list might be VERY slow if the size of this list is more than, say, one hundred. Since `contains` is called inside `NoteFilterModel::filterAcceptsRow`, in reality `contains` is called for each note within the note model - i.e. thousands of times. It is a classic example of an algorithm with quadratic complexity - it doesn't scale well; `m_noteLocalUids` should be `QSet<QString>` instead of `QStringList`.

**Comment by robert7** [comment]
Ok. If it is so simple, then fine.

I think you should get a database with few thousand notes to test (I guess you don't have one currently, as you would experience the problem yourself).

**Comment by d1vanov** [comment]
Yes, my own Evernote account is pretty minimalistic - I have only ~700 notes which is not as much as several thousands. It should be possible to autogenerate some test account with a script but it would only be useful as a local account, not Evernote one.

**Comment by d1vanov** [comment]
In fact, this slow notes filtering might also be the reason of slow notes loading into the editor - the GUI thread might be simply blocked by doing filtering thus preventing the note editor from doing its work on loading the note.

**Comment by robert7** [comment]
>It should be possible to autogenerate some test account 

Yes you can autogenerate. But you can temporarily join some public notebook (what ever content) and copy some notes (eventually without images). I did like this for one test account. With this process you would get more meaningful test data then by autogenerating. Or maybe import something from wikipedia via ENEX format.

**Comment by d1vanov** [comment]
Pushed a change to development which should improve the speed of notes filtering. Could you please try and tell if it makes any difference or not?

**Comment by robert7** [comment]
Unfortunately it is still slow (I don't have time measures (we should add them in log), but by feeling, I don't see any significant difference).

**Comment by d1vanov** [comment]
Ok, then we need to figure out where Quentier spends most of this long time. Logs are a poor tool to measure performance although I do have some ideas of which events we could log which would make it possible to judge how much time particular things take.

Could you please download and install [this tool](https://github.com/tbricks/tbstack) and use it to collect some stacks from Quentier process after you entered the search string and Quentier started to show some notes. You can just repeat the following command
```
tbstack <pid of Quentier process>
```
like 5-6 times, I hope that would be representative enough. Each run of this command would produce a separate file with the process stack - please share these with me here.

**Comment by robert7** [comment]
I tryied to build tbstack but I get
```
checking for library containing elf_begin... no
configure: error: unable to find the elf_begin() in libelf
```

**Comment by robert7** [comment]
Wouldn't it be easier if you try to repeat yourself by getting some more test data?
Here are some public notebooks with recipes :) https://www.google.sk/search?q=site:www.evernote.com/pub+recipe&filter=0&biw=1860&bih=885
You can join some of them and simply download like thousand few thousand notes (or get 1000 and duplicate 3x).

**Comment by d1vanov** [comment]
About building tbstack: I was able to do it after installing packages `libelf-dev`, `libunwind-dev` and `libdwarf-dev`. Also, I was a bit incorrect about how it works: it puts output onto stdout so need to redirect it to some file like
```
tbstack <pid of Quentier process> > stack1.txt
tbstack <pid of Quentier process> > stack2.txt
```

**Comment by robert7** [comment]
Ok. Seems I was able to compile the tbstack now (with your hint regarding packages). I can send you the data (later today), no problem.

**Comment by robert7** [comment]
I did 3 sessions - 1 capture per second - all 3 sessions should capture the problem (as the loading of one note takes like 10+ seconds). What I did: scroll down in the note list and randomly click on notes - so they did open in tabs.
During the 1st session did the sync run (at least the animation did run - I'm not sure, if sync did run really, as it taken quite long and it had only few notes to download). So I terminated the app, then in session2 and 3 the sync did not run.
 
[tbstack-captures.zip](https://github.com/d1vanov/quentier/files/2227767/tbstack-captures.zip)

**Comment by d1vanov** [comment]
Thanks! Will take a look into it.

**Comment by d1vanov** [comment]
Couldn't resist the temptation and looked through some of the stacks you captured. A lot of them capture `LocalStorageManagerPrivate::findAndSetResourcesPerNote` which is a part of the stack listing notes for `NoteModel`, the model serving the notes list from which you pick notes. So I went to take a look at that method and here's what I saw there:
```
bool LocalStorageManagerPrivate::findAndSetResourcesPerNote(Note & note, ErrorString & errorDescription,
                                                            const bool withBinaryData) const
{
    ErrorString errorPrefix(QT_TR_NOOP("can't find resources for note"));

    const QString noteLocalUid = note.localUid();

    // NOTE: it's weird but I can only get this query work as intended,
    // any more specific ones trying to pick the resource for note's local uid fail miserably.
    // I've just spent some hours of my life trying to figure out what the hell is going on here
    // but the best I was able to do is this. Please be very careful if you think you can do better here...
    QString queryString = QString(QStringLiteral("SELECT * FROM NoteResources"));
    QSqlQuery query(m_sqlDatabase);
    bool res = query.exec(queryString);
    DATABASE_CHECK_AND_SET_ERROR();
    <...>
}
```
`NoteResources` table contains **both** metadata **and** data for note's resources i.e. attachments. Typically things you attach to notes are much heavier by size than note content itself. `NoteModel` doesn't really need these resource bodies and it actually subscribes to notes with a special flag - `without resource binary data`, specifically to avoid extracting heavy blobs from the database. And here this little miserable query actually extracts these blobs from the database only to discard them in a moment. That's why notes loading is dog slow - because the database is busy extracting large blobs which do not go anywhere.

The comment near this query is also funny, isn't it? :blush: The time has come to deal with this nuisance.

**Comment by robert7** [comment]
OK. Let see.
Anyway the sync is also extremely slow. For 5 remotely changed notes and 2 locally changed it taken like 10 minutes or more. (I will look at it later in more detail).

**Comment by d1vanov** [comment]
It seems the slowness is caused by the initial loading of notes from local storage db into the in-memory note model. You can do the following experiment: apply this patch
```
diff --git a/src/models/NoteModel.cpp b/src/models/NoteModel.cpp
index 10772f5c..0c196f3b 100644
--- a/src/models/NoteModel.cpp
+++ b/src/models/NoteModel.cpp
@@ -1001,6 +1001,10 @@ void NoteModel::onListNotesComplete(LocalStorageManager::ListObjectsOptions flag

     m_listNotesRequestId = QUuid();

+    m_allNotesListed = true;
+    Q_EMIT notifyAllNotesListed();
+
+    /*
     if (!foundNotes.isEmpty()) {
         NMTRACE(QStringLiteral("The number of found notes is greater than zero, requesting more notes from the local storage"));
         m_listNotesOffset += static_cast<size_t>(foundNotes.size());
@@ -1009,6 +1013,7 @@ void NoteModel::onListNotesComplete(LocalStorageManager::ListObjectsOptions flag
     }

     checkAndNotifyAllNotesListed();
+    */
 }
```
and see whether the speed of note loading and sync improves - I believe it should. Actually I am already quite certain the slowness comes from this initial listing but such experiment should confirm it. It is only experiment but not a fix because e.g. counting of notes per tag currently relies on the completeness of the note model.

**Comment by d1vanov** [comment]
I believe I was finally able to grok the reason of slowness: `NoteResources` table contains mappings between notes and their resources via their unique ids. When we try to find the list of resources (even without their heavy binary contents but metadata only) for a particular note, `LocalStorageManagerPrivate::findAndSetResourcesPerNote` method is called. The db query run by this method
```
SELECT * FROM NoteResources
```
extracts mappings between notes and resources for **all** notes each time. And this method is called for **each** note separately during the initial listing of notes into the note model. If you have a lot of notes containing resources (and I guess that with several thousands of notes you have a large enough portion of notes containing attachments), this method would cause the quadratic complexity of notes listing. So this query should be fixed to only return resource ids for the required note, not for all notes.

**Comment by robert7** [comment]
can you pls commit in a feature branch? someway patch apply did not work (I never used patches before :) may be I did something wrong); but commiting in feature branch is no problem, even if it doesn't work; IMO patches are unnecessary hassle with git :)

![screen_20180725_05](https://user-images.githubusercontent.com/1135218/43207758-0bbf634c-9029-11e8-86a6-ce8ad82d59b4.png)

**Comment by d1vanov** [comment]
Ok, pushed it to branch `temporary/Issue-206-skip-most-part-of-note-model-filling-experiment`. Actually, I'm not quite sure it would help a great deal but: with this change you'd see only a tiny portion of your notes (100 to be precise) but hopefully rather shortly after the app launches notes should start opening quickly enough. You initially said it starts to work quickly after 10 minutes since the launch, this change should reduce this time to like 1 minute.

**Comment by robert7** [comment]
I compiled your branch and yes, then notes are now loading quickly.

**Comment by d1vanov** [comment]
Ok, thanks, I believe it confirms my suspicions of where the slowness comes from. Will try to improve it now.

**Comment by d1vanov** [comment]
Just pushed a change to libquentier/development which should speed up searching for resources per note. Will also look more closely to the stacks you collected to see if there's any other obvious enough bottleneck.

**Comment by d1vanov** [comment]
Hmm, enough stacks appear to contain `LocalStorageManagerPrivate::findEnResource` method which I did not optimize yet. It might be the next thing I'd need to take a closer look at but first please try the version I pushed, see if it at least slightly improves the situation - I believe it should.

**Comment by robert7** [comment]
I recompiled; if seems a bit faster (not sure as I have not time measurements - I try later add some).
But its still quite slow.
Here some logs from current version: 
[tbstack-logs.zip](https://github.com/d1vanov/quentier/files/2230875/tbstack-logs.zip)

**Comment by d1vanov** [comment]
Thanks! I skimmed quickly through about the half of the stacks you collected and about 80% of them capture `LocalStorageManagerPrivate::findEnResource` so I was right that it is the next candidate for optimization. I don't have the exact understanding why it is slow but I have a hypotheses and I know what can be tried to improve the situation here.

**Comment by d1vanov** [comment]
Pushed one more change which *might* make extracting the resource data faster. Before this commit I used a kind of a shortcut called [view](https://www.sqlite.org/lang_createview.html) in SQLite to list data from resources table but not including the heavy blobs i.e. the actual resource binary data. It might have lead to the use of a linear search within all resources' data performed by SQLite when processing the query. I wasn't able to fully verify it as SQLite told me it cannot explain me the plan for the query:
```
sqlite> EXPLAIN QUERY PLAN SELECT * FROM ResourcesWithoutBinaryData, LEFT OUTER JOIN ResourceAttributes ON ResourcesWithoutBinaryData.resourceLocalUid = ResourceAttributes.resourceLocalUid LEFT OUTER JOIN ResourceAttributesApplicationDataKeysOnly ON ResourcesWithoutBinaryData.resourceLocalUid = ResourceAttributesApplicationDataKeysOnly.resourceLocalUid LEFT OUTER JOIN ResourceAttributesApplicationDataFullMap ON ResourcesWithoutBinaryData.resourceLocalUid = ResourceAttributesApplicationDataFullMap.resourceLocalUid LEFT OUTER JOIN NoteResources ON ResourcesWithoutBinaryData.resourceLocalUid = NoteResources.localResource WHERE resourceGuid='11882db4-c6e2-48cf-9e6f-c425250b051f';
Error: RIGHT and FULL OUTER JOINs are not currently supported
```
So I "rephrased" this query in a more explicit way, through listing all the exact columns I want from all involved tables without using the view shortcut:
```
sqlite> EXPLAIN QUERY PLAN SELECT Resources.resourceLocalUid, resourceGuid, noteGuid, resourceUpdateSequenceNumber, resourceIsDirty, dataSize, dataHash, mime, width, height, recognitionDataSize, recognitionDataHash, alternateDataSize, alternateDataHash, resourceIndexInNote, resourceSourceURL, timestamp, resourceLatitude, resourceLongitude, resourceAltitude, cameraMake, cameraModel, clientWillIndex, fileName, attachment, resourceKey, resourceMapKey, resourceValue, localNote FROM Resources LEFT OUTER JOIN ResourceAttributes ON Resources.resourceLocalUid = ResourceAttributes.resourceLocalUid LEFT OUTER JOIN ResourceAttributesApplicationDataKeysOnly ON Resources.resourceLocalUid = ResourceAttributesApplicationDataKeysOnly.resourceLocalUid LEFT OUTER JOIN ResourceAttributesApplicationDataFullMap ON Resources.resourceLocalUid = ResourceAttributesApplicationDataFullMap.resourceLocalUid LEFT OUTER JOIN NoteResources ON Resources.resourceLocalUid = NoteResources.localResource WHERE resourceGuid='11882db4-c6e2-48cf-9e6f-c425250b051f';
0|0|0|SEARCH TABLE Resources USING INDEX sqlite_autoindex_Resources_2 (resourceGuid=?)
0|1|1|SEARCH TABLE ResourceAttributes USING INDEX sqlite_autoindex_ResourceAttributes_1 (resourceLocalUid=?)
0|2|2|SEARCH TABLE ResourceAttributesApplicationDataKeysOnly USING COVERING INDEX sqlite_autoindex_ResourceAttributesApplicationDataKeysOnly_1 (resourceLocalUid=?)
0|3|3|SEARCH TABLE ResourceAttributesApplicationDataFullMap USING INDEX sqlite_autoindex_ResourceAttributesApplicationDataFullMap_1 (resourceLocalUid=?)
0|4|4|SEARCH TABLE NoteResources USING AUTOMATIC COVERING INDEX (localResource=?)
```
From this plan it seems no linear search would be involved into the processing of such query. So, please try this version.

**Comment by robert7** [comment]
Seems to be faster (like from 10-15 seconds down to 5-7 seconds per one note). Also much better, but still quite slow.
I will try to all some logs with time measurements.

I think you should do some kind of "lazy loading". Only load the resources, then opening one particular note and only do this for the one note loaded (maybe you are already doing this). So the time to load one note with resources should not much rely on count of notes in database.
I try to look at the data model during the weekend.

**Comment by d1vanov** [comment]
Thanks for the feedback! May I ask you to collect some stacks once again? Curious to see where the bottleneck might be this time - is it still the same place or something else. Also, do you still observe the effect of note loading getting faster after some time since the app's startup?

If by "lazy loading" you mean not loading all notes from the database on startup then yes, it might be considered although it brings in quite a lot of complications regarding notes sorting and filtering. Currently notes are loaded from the db into the model in chunks of 100 notes per slot invokation. The binary data of the resources is not included into these chunks so their huge size should not interfere and cause any slowness.

**Comment by d1vanov** [comment]
In fact, I can already see some other potentially slow queries within the local storage: first, the invocation of `LocalStorageManager::findNote` with `withResourceBinaryData` parameter set to false presumably has precisely the same problem with SQLite view working slowly. Furthermore, turned out I forgot to add an index into one more supplementary table even if we want to get the note with resource binary data using guid (as opposed to local uid) as the search key:
```
sqlite> EXPLAIN QUERY PLAN SELECT * FROM Notes LEFT OUTER JOIN SharedNotes ON ((Notes.guid IS NOT NULL) AND (Notes.guid = SharedNotes.sharedNoteNoteGuid)) LEFT OUTER JOIN NoteRestrictions ON Notes.localUid = NoteRestrictions.noteLocalUid LEFT OUTER JOIN NoteLimits ON Notes.localUid = NoteLimits.noteLocalUid LEFT OUTER JOIN Resources ON Notes.guid = Resources.noteGuid LEFT OUTER JOIN ResourceAttributes ON Resources.resourceLocalUid = ResourceAttributes.resourceLocalUid LEFT OUTER JOIN ResourceAttributesApplicationDataKeysOnly ON Resources.resourceLocalUid = ResourceAttributesApplicationDataKeysOnly.resourceLocalUid LEFT OUTER JOIN ResourceAttributesApplicationDataFullMap ON Resources.resourceLocalUid = ResourceAttributesApplicationDataFullMap.resourceLocalUid LEFT OUTER JOIN NoteTags ON Notes.localUid = NoteTags.localNote WHERE guid='001a82df-d7a2-4afd-95a0-c6041b265c28';
0|0|0|SEARCH TABLE Notes USING INDEX sqlite_autoindex_Notes_2 (guid=?)
0|1|1|SEARCH TABLE SharedNotes USING INDEX sqlite_autoindex_SharedNotes_2 (sharedNoteNoteGuid=?)
0|2|2|SCAN TABLE NoteRestrictions
0|3|3|SEARCH TABLE NoteLimits USING AUTOMATIC COVERING INDEX (noteLocalUid=?)
0|4|4|SEARCH TABLE Resources USING AUTOMATIC COVERING INDEX (noteGuid=?)
0|5|5|SEARCH TABLE ResourceAttributes USING INDEX sqlite_autoindex_ResourceAttributes_1 (resourceLocalUid=?)
0|6|6|SEARCH TABLE ResourceAttributesApplicationDataKeysOnly USING COVERING INDEX sqlite_autoindex_ResourceAttributesApplicationDataKeysOnly_1 (resourceLocalUid=?)
0|7|7|SEARCH TABLE ResourceAttributesApplicationDataFullMap USING INDEX sqlite_autoindex_ResourceAttributesApplicationDataFullMap_1 (resourceLocalUid=?)
0|8|8|SEARCH TABLE NoteTags USING INDEX NoteTagsNote (localNote=?)
```
The important part from the above is
```
0|2|2|SCAN TABLE NoteRestrictions
```
This one is not indexed i.e. it performs linear search over all notes. It is a little unexpected since I left outer join this table on note local uid even if we search the note itself by guid. Perhaps I need to insert the guid column into this table and see if it would help.

In addition to this I need to get rid of this SQLite view thing as, given the latest feedback, it does perform much slower than the original table.

**Comment by robert7** [comment]
here are the stacks from today [tbstacks.zip](https://github.com/d1vanov/quentier/files/2235656/tbstacks.zip)

I will look at the data model during the weekend and maybe then I can give you more useful feedback

**Comment by d1vanov** [comment]
Ok, most of the stacks are still within `LocalStorageManagerPrivate::findEnResource`'s SQL query. As I read [here](http://sqlite.1065341.n5.nabble.com/slow-select-from-table-with-data-blob-td44396.html), it is a known "feature/issue" of SQLite - selection from tables containing large enough blobs, even if no blobs are actually selected, is slow. The recommended way to handle it is to either not include the large blobs into the database or put them into separate tables and not touch these tables unless blobs are required. It seems the latter option is the one I should implement, only it would require updating the database schema and the database upgrade procedure. It would take me a while to implement this.

**Comment by robert7** [comment]
I added some INFO messages with simple time logging and did some measurements (its committed in feature branch "time-measure" in case you wanna look). The measurement are not precise, but I think they give good picture.

1. loading of all notes on my machine takes 7,6 minutes (464 seconds)
2. after all notes are loaded, displaying of individual notes is instant
3. the SQL command for extracting all notes with the joins you do (SELECT * FROM Notes ...) is pretty fast (like 250 ms) - see screenshoot from sqlitebrowser - slow part is loading resources LocalStorageManagerPrivate::listNotes() (which actually calls LocalStorageManagerPrivate::findEnResource() as you identified correctly)

Here is INFO log from my real note account
[Quentier-log.zip](https://github.com/d1vanov/quentier/files/2236749/Quentier-log.zip)
finish loading `2018-07-27 19:45:24.503 CEST src/models/NoteModel.cpp:2209 [Info]: [non-deleted notes] All the necessary data items were listed; total time = 464822 ms`

Loading resources for 100 notes takes on average 4000 ms `2018-07-27 19:45:14.723 CEST libquentier/src/local_storage/LocalStorageManager_p.cpp:2403 [Info]: LocalStorageManagerPrivate::listNotes: flag = QFlags(), withResourceBinaryData = false, linked notebook guid = , returning 100 in 5655 ms`

OK what I actually don't understand and what I previously meant with "lazy loading" is, that I think, that for the "NoteListView" you actually only need result from "SELECT * FROM Notes" (I would even not do "*" but only select relevant colums like title, last modified, date created - only the column you need for  "NoteListView" - displaying list of notes and sorting them.

ALL other data (like thumbnail data, content text of note for preview) I would only load for notes which are actually displayed. I don't think you can load resources for 4K notes in 1 second - but you can load all the metadata + resources for the notes which are currently in viewport - "lazy load". The images etc. I would only load for the particular note which is about to be opened in editor.

![screen_20180727_03](https://user-images.githubusercontent.com/1135218/43339230-13fb353e-91d9-11e8-893d-a2bf732b5d5a.png)

**Comment by d1vanov** [comment]
> OK what I actually don't understand - why you are loading resources at all?

It's true the full data about resources is not required. However, it might be useful for note list view (and hence the note model serving it) to at least know whether the note contains any resources at all - for such notes something like a clip icon could be rendered. I never wanted to load the actual heavy binary data of these resources and I do not in fact do that. But it turns out extracting anything from a table which actually contains heavy blobs is almost as slow as if I did extract these heavy blobs too, that's the problem.

Although it might be possible to do a short term fix of not loading even the metadata about the resources, I wouldn't really like to do it. Instead I'd like to fix the problem with slow loading of this metadata. If you'd like, you can probably do a short term fix for your fork yourself - just don't load resources at all in `LocalStorageManagerPrivate::listNotes` if `withResourceBinaryData` is false. I'm not sure it won't break anything else though (like sync), some testing would be required.

> ALL other data (like thumbnail data, content text of note for preview) I would only load for notes which are actually displayed.

Well, it is possible but it sufficiently complicated things. The communication between the note model and local storage is asynchronous, via signals and slots while the communication between the model and the view is synchronous. So the only viable scheme is like what I did in #143 - LRU cache of "full" data entries which are delivered asynchronously if/when required but it's quite a complex thing to implement + note model has other complications, for example, tag model uses it to count the numbers of notes per tag. So I think it's really simpler to fix the slow query rather than to do this complex redesign of the note model.

**Comment by d1vanov** [comment]
As I mentioned previously, the approach I plan to take on this issue involves the change of the database schema, so the db upgrade procedure needs to be implemented, tested and made user-friendly: user should be warned that his/her local storage database is going to be upgraded and the previous version of the app won't be able to work with it any longer. There should be a possibility to opt out of this db upgrade. All these little things together involve a lot of coding of different things. Unfortunately, I'm quite busy with personal matters this weekend + I'm going to be on a mini-vacation mid next week - next next week. So this issue most likely would take at least three more weeks - just a rough ETA.

**Comment by robert7** [comment]
OK. Just I would not over complicate with the "opt-out" (in my case I would just write info in release notes - and maybe a warning dialog box about making backup before running new version).

And another idea for speedup: if I understand correctly you do 1 or more SQL calls per one node to get the resource info. There is time overhead with each SQL query. If would be faster, if you could request the resource infos for the 100 note block with one SQL command. This should be possible and could be possibly faster even without changing db schema.

Just out of curiosity: do you know how roughly how many active users of the current version are "out there"?

**Comment by d1vanov** [comment]
Thanks for the idea, it should indeed be possible to extract resource data for a bunch of notes with a single query. I think it makes sense to consider it in future as an optimization measure. However, I don't really believe it would help with the current issue since, as I understand, SQLite spends most of the time skimming through disk pages containing blob data, not compiling multiple queries.

I don't know how many active users there are right now. Probably not many. But it doesn't mean I shouldn't treat them with respect - too often the software I myself am using does breaking and backward-incompatible changes behind my back and doesn't leave me any options to consider. I wouldn't like to repeat this pattern, especially in free software which is supposed to respect its users.

**Comment by robert7** [comment]
>But it doesn't mean I shouldn't treat them with respect..

Yes. But improving faster is also paying respect to future users (which may be much more in count the current alpha version user count). So if we get from alpha stage faster, this may do more good in absolute value. So I don't say ignore present users, just be pragmatic.

**Comment by robert7** [comment]
So I don't say ignore existing users or being respect-less, but being rational and something in direction of what's called "[effective altruism](https://en.wikipedia.org/wiki/Effective_altruism)". But this is just my 5 cents, do what you think its best, any improvement is appreciated!

**Comment by d1vanov** [comment]
In fact, you might be right. The shorter path actually exists in this case - just change the API for notes listing to allow not touching resources at all in queries. For the current state of note model it should be enough. No database schema upgrade would be required. I'll think about it.

**Comment by d1vanov** [comment]
I decided to do the short term fix: introduce the possibility to not touch the resources table at all while listing notes and use it in note model. Can't say I like this option but other ones seem to require quite a lot of time and effort.

**Comment by robert7** [comment]
I think its good idea! There is always possibility to improve later.

**Comment by robert7** [comment]
I would try to do following: load all notes without resource info - add variable to model indicating that for particular note the resource info has not been yet loaded (like "resourceLoaded = false").
Then on event, where particular list of notes comes to view and for particular note resourceLoaded==false, send signal "resourceInfoLoadRequested(noteLocalUid)". 
NoteListView could be good place.

Then resource info will be loaded async, but only for notes, which are currently displayed - which will be very fast. On slot handling the mentioned signal, load the resource data and set resourceLoaded=true and redraw the note in NoteListView.

**Comment by d1vanov** [comment]
Right now `NoteListView` does not in fact require the information whether particular notes have resources or not. For notes which have thumbnails it draws them. There is a check currently in `NoteItemDelegate` which only draws the thumbnail if note contains resources but in most cases it is redundant since it the thumbnail data is present, the note at least had resources in the past and most probably still has them. So this check can be removed for now.

As I see it, the right way to go should be doing what I originally planned to do here: change the database schema to make loading resources metadata (without the actual heavy binary data) fast enough for the purposes of `NoteListView` and then enable the retrieval of this metadata by `NoteModel`. I still plan to actually do that in some observable future after the introduction of the short term fix.

And even if it was actual to load resource data asynchronously for notes which require it, I don't really agree `NoteListView` is a good place for this: the view should rely on the data provided by the model and only care about the data display via the delegate (`NoteItemDelegate` in this case) and about handling various actions on the data items. In the ideal case it should be the model which subscribes to stuff asynchronously when it's requested and only maintains a small subset of data which might become required. Take a look at `LogViewerModel` in development, after #143 it represents exactly what I'm talking about here.

**Comment by d1vanov** [comment]
Pushed the changes for both libquentier and Quentier to development - now NoteModel would request notes list in a way not involving touching the resources table at all. Please rebuild and give it a try, it should work faster now.

**Comment by robert7** [comment]
OK. i'll check

**Comment by robert7** [comment]
I compiled the current version. Initial load is not much more faster down to ~65 seconds from 7.6 minutes.
But unfortunately opening some random notes still takes ages (it seems it even got worse). I measured with stopwatch on my mobile - opening of one note sometimes taken 30-48 seconds (I even got one with 59 seconds). Some open quickly, but some take very very long time. Also the sync is extremely slow - like I changed like 15 notes on remote and the sync is already running like 15 minutes (at least the animation is still running; not sure what it does).

Opening one note:
![2018-07-31 22 18 32](https://user-images.githubusercontent.com/1135218/43484979-364f35ce-9510-11e8-8dd2-1f37eb830753.png)
![2018-07-31 22 19 52](https://user-images.githubusercontent.com/1135218/43484981-366dabf8-9510-11e8-9b5a-0eb6361d1a02.png)

I really tried to use quentier for like 1.5 days for real work, but it is not yet ready. So I "recomputed the priorities" :) and temporarily decided to fix the most urgent bugs in my version of nixnote and stay a while with it until quentier gets more stable. It is a bit paradox, that although quentier has much better architecture and cleaner code, nixnote is actually more stable. And basically works good enough for my needs.

I can continue help with testing. For the performance issue it could be useful if you get some more test data - like 5000+ notes - e.g. from the public Evernote notebooks (I already posted a link).

**Comment by d1vanov** [comment]
> Initial load is not much more faster down to ~65 seconds from 7.6 minutes

Well, although it still might be not as fast as desired, it's almost 7x improvement actually :smiley:

> Some open quickly, but some take very very long time

I'm curious if notes opening so slowly contain attachments.

> Also the sync is extremely slow - like I changed like 15 notes on remote and the sync is already running like 15 minutes (at least the animation is still running; not sure what it does).

Well, it could have gotten stuck due to some mistake in the sync logic. Or maybe it really takes that much time to update some notes with full resource binary data, can't say for sure without logs.

> I really tried to use quentier for like 1.5 days for real work, but it is not yet ready. So I "recomputed the priorities" :) and temporarily decided to fix the most urgent bugs in my version of nixnote and stay a while with it until quentier gets more stable. It is a bit paradox, that although quentier has much better architecture and cleaner code, nixnote is actually more stable. And basically works good enough for my needs.

I don't think there's a paradox here - Quentier has a much more complicated architecture and it is always harder to get such one right than something more primitive. Plus, if you count Nixnote 1, Nixnote has been around for almost a decade now while Quentier is quite a young project so far. Ok, I understand your motivation but Quentier won't get stable any faster without your help rather that with it :wink: 

> I can continue help with testing. For the performance issue it could be useful if you get some more test data - like 5000+ notes - e.g. from the public Evernote notebooks (I already posted a link).

Ok, I will try to set things up on my side. However, if it won't be too hard for you, stacks from loading some of the slowest notes could help to understand where the bottleneck is (I suppose it's still in loading resources but would like to be sure).

**Comment by d1vanov** [comment]
So I'll continue with #213 and d1vanov/libquentier#69 as I planned before.

**Comment by robert7** [comment]
I'' collect the stacks for you. I did few more testing today and most of the note load really quickly but some take **really** long like few tens of seconds.

>Perhaps I should embrace Nixnote's model of resources storage

Actually I think storing resources as blobs is better design. And it could be little bit slower, but it should **not** be slower by some orders of magnitude. So there must be some problem other that just usage of the BLOBs alone.

**Comment by d1vanov** [comment]
Stacks should help to determine where most of the time is spent while loading the particularly slowly loaded notes. I'd bet is on SQLite extracting resources but let's see.

> Actually I think storing resources as blobs is better design

I thought so as well but right now I doubt it was a good decision after all - I should have learned the limitations and intricacies of SQLite before designing the database schema.

**Comment by d1vanov** [comment]
> So there must be some problem other that just usage of the BLOBs alone.

I can think of one such problem: currently during loading into the editor each note is in fact extracted twice from the local storage: one time with resources metadata only (which we know now as nearly as slow as extracting it together with resource binary data) and one other time with full resources data. This scheme is kinda over-engineered and it can and should be reduced to only one load per opened note. But furthermore, if resources were stored as plain files, it might be possible to implement the note editor without loading the binary data of resources at all.

**Comment by robert7** [comment]
> I don't think there's a paradox here - Quentier has a much more complicated architecture and it is always harder to get such one right than something more primitive.  .... Nixnote has been around for almost a decade now while Quentier is quite a young project so far. 

> ... However, it's a pretty large amount of work which would most likely take several months.

I don't think you need many months to get really stable version of the app. Without being stable, you don't get much real users. And without real users, then project is a bit "academic" (IMO). And you don't really know how much time you have ... maybe in half a year Evernote will finally wake up and dramatically improve their Web Client and any desktop client will be pointless for most users. Or Microsoft will add easy import of Evernote data into Onenote and people will leave Evernote.. or Google will improve Keep..

Here are some measurements. I did first wait ~65 seconds until model is loaded.
Then the slowness was not apparent from beginning, but it seems that it began after I started Sync at around 08:28:30, then the slowness was apparent at around 08:30:00 and later (see timestamp in filenames). The sync taken very long, I ended program before it finished. There weren't many notes to sync.

[tbstacks-and-log.zip](https://github.com/d1vanov/quentier/files/2252357/tbstacks-and-log.zip)

**Comment by d1vanov** [comment]
Thanks for the stacks, will look into them later.

About the time speculations - I certainly don't know how much time I have. You know, there's always that bus lurking around the corner. But, I don't honestly have conquering the world of Evernote users or even the world of Evernote & Linux users as my top priority. So yes, this project is admittedly a bit academic and I'm fine with it. I'm working on this project because

1. I'm actually enjoying the process. It's nice to be free to design things as I see fit and to be free to use whatever technologies I want as opposed to my money making job.
2. I want to (eventually) have something looking good enough as a portfolio project.
3. Because I need Evernote client with the possibility for offline note taking for my own needs. 
4. Because it might at some point become useful for Evernote users on Linux and beyond.

So if the landscape changes before the app gets stable, well, it's a pity but I'll live with it, I guess. Maybe I'll think of using some other backend than Evernote - I already had requests to be able to use git repository for notes storage. It is actually something another app called [MyTetra](https://github.com/xintrea/mytetra_dev) does. Or maybe I'll just move on to some other project I have in mind. Let's see.

**Comment by d1vanov** [comment]
Looked through some of the stacks - now they seem dominated by updating notes and resources within the local storage. Updating resources is understandably slow because SQLite is slow to work with large blobs and updating notes is most likely slow because it checks whether resources need to be updated and for this queries the information from resources table which is slow as we know now. So, I did not see anything unexpected so far. #213 and d1vanov/libquentier#69 are the way to go.

**Comment by robert7** [comment]
> But, I don't honestly have conquering the world of Evernote users or even the world of Evernote & Linux users as my top priority.

Of course. Me neither! :) Just with real users is much more fun.
But OK.

**Comment by d1vanov** [comment]
It took me almost half a year but I implemented the planned change of the db schema - binary data of resources is now stored in plain files, outside the primary db. This data is not touched for notes listing and only image resources are loaded into memory when any particular note is loaded into the editor. The changes are in development branches of libquentier and quentier.

I'm still not done here - I plan to look into making the note model not load the info on all notes on startup. It seems feasible even though relatively complex.

**Comment by d1vanov** [comment]
New note model loading notes lazily has landed in development branch. Notes listing on startup should now feel much better and should not get in the way of loading notes into note editor pages. Also listing of notes corresponding to search queries should feel much faster as filtering is now built into the note model itself. Previously filtering was a separate layer which didn't play very nice with large amounts of notes per account.

I'd like to take one more step further here and implement test account data generator so that I can test performance with truly large accounts. So keeping this one open for now.

**Comment by d1vanov** [comment]
This one took me a really long time overall - almost a year. I had to introduce several changes on different levels of the application - change the storage format for attachment data + implement the migration procedure, rewrite the note model in a way to get rid of loading metadata on all notes within the account into memory + remove slow and unreliable notes filtering layer. Finally, in #237 I implemented a tool to load random wiki articles and convert them into notes, so I was finally able to compose a large enough account on which I can test the performance of notes loading. It seems good enough to me now - on an account consisting of 10000 notes containing collectively about 13000 attachments I can see that the slowest load time is about a second which seems bearable. Additionally I made the note editor display "Loading note..." text while the note page is being prepared so it's not as confusing now as it used to be.

All these changes are now in development branch. The difference between master and development is quite large now so in the next couple of months I'll be focusing on getting the development stable enough to be merged into master.
